_posts/pl/2023-09-08-czy-wiesz-po-co-stosuje-sie-sneaky-throws-z-biblioteki-lombok.md

```diff
@@ -0,0 +1,145 @@
+      File file = new File("Nieistniejacy_plik.txt");
+      try {
+          FileInputStream stream = new FileInputStream(file);
+      } catch (FileNotFoundException e) {
+          e.printStackTrace();
+      }
+  }
+  ```
+
+## Koncepcja sneaky throws
+
+`Sneaky throws` to koncepcja pozwalająca na rzucenie dowolnego `checked exception` bez jego jawnego definiowania w sygnaturze metody. Pozwala ona na ominięcie słowa kluczowego throws oraz imitowanie zachowania `unchecked exception` i jest możliwa, ponieważ obsługa wyjątków `checked exception` jest wymuszana tylko przez kompilator Javy.  W kodzie bajtowym każdy wyjątek może zostać rzucony z dowolnego miejsca i jest traktowany przez JVM tak samo - zostaje on propagowany w dół stosu wywołań. Od Javy 8 każde użycie throws `T`, gdzie `T` jest typem generycznym rozszerzajacym `Throwable`, oznacza, że metoda może rzucić `unchecked exception`. Dzięki temu możemy stworzyć metodę pomocniczą, która będzie rzucała wyjątek typu `checked`, jednak kompilator nie będzie wymagał jego przechwycenia.
+Metoda pomocnicza realizująca koncepcję `sneaky throws`:
+```java
+private static <T extends Throwable> void sneakyThrow(Throwable t) throws T {
+    throw (T) t;
+}
+```
+
+Metodę tę możemy następnie dowolnie wykorzystać w naszym kodzie:
+```java
+private File getFile(String fileName) {
+   return null;
+}
+ 
+private void deleteFile(String fileName) {
+   File file = getFile(fileName);
+   if (file == null) {
+      sneakyThrow(new FileNotFoundException("Nie znaleziono pliku"));
+   }
+   file.delete();
+}
+ 
+public void tryToDeleteNotExistingFile() {
+   try {
+      deleteFile("Nieistniejacy_plik.txt");
+   } catch (Exception exception) {
+      exception.printStackTrace();
+   }
+}
+```
+Warto zauważyć, że w takiej sytuacji w metodzie `tryToDeleteNotExistingFile()` nie możemy złapać już wyjątku `FileNotFoundException`, ponieważ nie jest on zadeklarowany - możemy jedynie ratować się złapaniem bardziej ogólnego `Exception`.
+
+## Lombok - adnotacja `@SneakyThrows`
+Biblioteka `Lombok` udostępnia adnotację `@SneakyThrows`, która wykorzystuje powyższą sztuczkę i dzięki oszukaniu kompilatora pozwala rzucać `checked exception` bez deklarowania tego w sygnaturze metody.
+Przykład - użycie adnotacji `@SneakyThrows`:
+```java
+private File getFile(String fileName) {
+   return null;
+}
+ 
+@SneakyThrows(FileNotFoundException.class)
+private void deleteFile(String fileName) {
+   File file = getFile(fileName);
+   if (file == null) {
+      throw new FileNotFoundException("Nie znaleziono pliku");
+   }
+   file.delete();
+}
+ 
+public void tryToDeleteNotExistingFile() {
+   try {
+      deleteFile("Nieistniejacy_plik.txt");
+   } catch (Exception exception) {
+      exception.printStackTrace();
+   }
+}
+```
+Do adnotacji `@SneakyThrows` można przekazać dowolną liczbę wyjątków. Jeśli nie podamy żadnego to adnotacja ta uwzględni dowolny wyjątek. Należy pamiętać także o tym, że `@SneakyThrows` nie dziedziczy.
```

Comment:
'Jeśli nie podamy żadnego to adnotacja ta uwzględni dowolny wyjątek.' - dodajmy przecinek po 'żadnego'